client.go

```diff
@@ -189,34 +189,100 @@ func (c *Client) exchange(m *Msg, a string) (r *Msg, rtt time.Duration, err erro
 // If the received message contains a TSIG record the transaction
 // signature is verified.
 func (co *Conn) ReadMsg() (*Msg, error) {
-	var p []byte
+	p, err := co.ReadMsgBytes(nil)
+	if err != nil {
+		return nil, err
+	}
+
 	m := new(Msg)
-	if _, ok := co.Conn.(*net.TCPConn); ok {
-		p = make([]byte, MaxMsgSize)
+	if err := m.Unpack(p); err != nil {
+		return nil, err
+	}
+	if t := m.IsTsig(); t != nil {
+		if _, ok := co.TsigSecret[t.Hdr.Name]; !ok {
+			return m, ErrSecret
+		}
+		// Need to work on the original message p, as that was used to calculate the tsig.
+		err = TsigVerify(p, co.TsigSecret[t.Hdr.Name], co.tsigRequestMAC, false)
+	}
+	return m, err
+}
+
+// ReadMsgBytes reads DNS packet, parses and fills wire header (passing nil
+// would cause skipping that). Returns message in byte format to parse with
+// Msg.Unpack later on.
+//
+// Note that this function would not be able to report TSIG error or
+// check it got actual DNS payload.
+func (co *Conn) ReadMsgBytes(hdr *Header) ([]byte, error) {
+	var (
+		p   []byte
+		n   int
+		err error
+	)
+
+	if t, ok := co.Conn.(*net.TCPConn); ok {
+		// we got two byte header to know how much to receive...
+		l, err := tcpMsgLen(t)
+		if err != nil {
+			return nil, err
+		}
+		p = make([]byte, l)
+		n, err = tcpRead(t, p)
 	} else {
 		if co.UDPSize > MinMsgSize {
 			p = make([]byte, co.UDPSize)
 		} else {
 			p = make([]byte, MinMsgSize)
 		}
+		n, err = co.Read(p)
 	}
-	n, err := co.Read(p)
-	if err != nil && n == 0 {
+
+	if err != nil {
 		return nil, err
+	} else if n < _HBytes {
+		return nil, ErrShortRead
 	}
+
 	p = p[:n]
-	if err := m.Unpack(p); err != nil {
-		return nil, err
+	if hdr != nil {
+		if _, err = UnpackStruct(hdr, p, 0); err != nil {
+			return nil, err
+		}
 	}
-	co.rtt = time.Since(co.t)
-	if t := m.IsTsig(); t != nil {
-		if _, ok := co.TsigSecret[t.Hdr.Name]; !ok {
-			return m, ErrSecret
+	return p, err
+}
+
+// tcpMsgLen - helper func to read first two bytes of stream as uint16 packet length
+func tcpMsgLen(t *net.TCPConn) (int, error) {
+	p := [2]byte{0, 0}
+	n, err := t.Read(p[:])
```

Comment:
I wanted to do static array, not slice here to force it being on stack. To turn it into a slice you need [:]... So I can do that probably anyway by removing 2 from [2] up top... I truly don't see any performance effect.